### PR TITLE
[codex] Add CUDA 12.2 compatibility fallbacks

### PR DESCRIPTION
**Pull request by BBuf**
## Summary
- Add compatibility fallbacks so current `main` can import, build `sgl-kernel` from source, and run on the CUDA 12.2 / PyTorch 2.1 stack shipped by `nvcr.io/nvidia/pytorch:23.10-py3`.
- Fix old-Torch custom-op ABI issues by converting `std::optional`/tuple returns through a PyTorch-compatible shim and guarding newer dispatcher APIs.
- Gate CUDA components that require newer headers/toolkits: FlashMLA and green-context `spatial_ops` require CUDA >= 12.4, and SM100 MXFP8 sources require CUDA >= 12.8.
- Keep accelerated AOT `sgl_kernel` common ops available on H100/CUDA 12.2 while preserving newer paths when their dependencies are present.

## Base
- Upstream main SHA: `d8a4f7a7aa235601322dc3f861c936ddd197a0cf`
- Patch commit: `f9899f5136bedac92407e0296dde4f3d11ade688`

## CUDA 12.2 Environment
- Host: H100 80GB
- Container image: `nvcr.io/nvidia/pytorch:23.10-py3`
- Python: 3.10.12
- Torch: `2.1.0a0+32f93b1`
- `torch.version.cuda`: `12.2`
- NVCC: `cuda_12.2.r12.2/compiler.33191640_0`
- Built wheel: `sglang_kernel-0.4.4-cp310-abi3-linux_x86_64.whl`

## Validation
Static checks:
```bash
git diff --check origin/main
```

AOT wheel build in CUDA 12.2 container:
```bash
cd /data/bbuf/repos/sglang-cuda122-main/sgl-kernel
TORCH_CUDA_ARCH_LIST=9.0 \
CMAKE_POLICY_VERSION_MINIMUM=3.5 \
CMAKE_ARGS="-DSGL_KERNEL_COMPILE_THREADS=1 -DENABLE_BELOW_SM90=OFF -DSGL_KERNEL_ENABLE_FA3=OFF" \
CMAKE_BUILD_PARALLEL_LEVEL=8 MAX_JOBS=8 \
uv build --wheel -Cbuild-dir=build_cuda122 . --verbose --color=never --no-build-isolation
python -m pip install --force-reinstall --no-deps dist/*.whl
```
Observed:
- CMake detected `CUDA_VERSION=12.2`.
- `spatial_ops` skipped because CUDA green-context APIs require CUDA >= 12.4.
- `flashmla_ops` skipped because FlashMLA requires CUDA >= 12.4.
- `sm90/common_ops.abi3.so` and `sm100/common_ops.abi3.so` linked successfully.
- Wheel built successfully.

AOT import and op smoke:
```bash
import torch, sglang, sgl_kernel
print(torch.__version__, torch.version.cuda)
print(sgl_kernel.common_ops.__file__)
```
Observed:
- `torch 2.1.0a0+32f93b1 cuda 12.2`
- `sgl_kernel 0.4.4 /usr/local/lib/python3.10/dist-packages/sgl_kernel/sm90/common_ops.abi3.so`
- Direct CUDA op correctness smoke passed for `rmsnorm`, `silu_and_mul`, and `top_k_renorm_probs` with both `None` and tensor optional-arg paths.
- A single upstream pytest case passed after adding a process-local Torch 2.1 test shim for missing `torch.get_default_device`:
  `sgl-kernel/tests/test_activation.py::test_fused_silu_mul[1-1-128]`.

Qwen2.5-32B model smoke on final head:
```bash
CUDA_VISIBLE_DEVICES=0 HF_HUB_DISABLE_XET=1 PYTHONUNBUFFERED=1 \
python -m sglang.launch_server \
  --model-path Qwen/Qwen2.5-32B-Instruct \
  --host 127.0.0.1 --port 31023 \
  --dtype bfloat16 \
  --attention-backend torch_native \
  --sampling-backend pytorch \
  --grammar-backend none \
  --disable-cuda-graph \
  --disable-radix-cache \
  --context-length 2048 \
  --mem-fraction-static 0.82 \
  --trust-remote-code
```
Observed:
- Local HF snapshot loaded; no model download needed.
- `Load weight end. elapsed=8.71 s, type=Qwen2ForCausalLM, mem usage=61.04 GB`.
- KV cache allocated successfully.
- `/health_generate` returned OK.
- `/generate` returned a normal Chinese response for `用一句话解释 CUDA 12.2 兼容性验证。`, with `completion_tokens=48`, `e2e_latency=1.65s`.
- The scheduler process mapped `/usr/local/lib/python3.10/dist-packages/sgl_kernel/sm90/common_ops.abi3.so`, confirming the AOT common ops path was loaded during the model smoke.
- The validation server was stopped after the smoke; GPU0 was released.

## Notes
- `--sampling-backend pytorch` is used because this clean CUDA 12.2 image does not include the FlashInfer Python package. This PR validates the AOT `sgl_kernel` common-ops path, not FlashInfer sampling.
- `--attention-backend torch_native` is used as the CUDA 12.2-compatible attention path for this environment.
- Full `sgl-kernel/tests/test_sampling.py` collection is blocked in this container by missing `flashinfer.sampling`; the test fixture also expects newer `torch.get_default_device`. Direct AOT op smoke covers the CUDA kernels and old-Torch optional-arg ABI paths relevant to this fix.

---
### CI States

Latest PR Test (Base): :x: [Run #28609423903](https://github.com/sgl-project/sglang/actions/runs/28609423903)
Latest PR Test (Extra): :x: [Run #28609423532](https://github.com/sgl-project/sglang/actions/runs/28609423532)

### CODE REVIEW

**Comment by BBuf** [comment]
Superseded by #30020, which contains the smaller CUDA 12.2 patch validated on the latest main.
